### DISCUSSION TITLE
Examples of reading SDF

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#699] [General]
EPOCH has 3 options to read proprietary SDF files - MATLAB, Python and some other proprietary app like PlPlot or something . Somebody for all that had to create special libraries how to deal with that SDF black box. But because EPOCH is compiled in Fortran there are no need to write any additional libraries for Fortran uses, they are already automatically written and successfully do all the outputs. 
  My question or request with this respect is --  can developers show any Fortran examples how to read SDF files and extract from them major data - i.e. particles and fields data ? Definitely, when developers initially wrote EPOCH's I/O facilities they  had to debug them and no other means of reading SDF files besides Fortran existed yet. 

Similarly, some easy guide and demo examples of reading SDF files with Python would be also appreciated. If Python users  also share their experience that would be great too. For example how fast is Python in reading and writing processed data ?

**Comment by TomGoffrey** [comment]
EPOCH can read SDF data. I would suggest starting in the `restart_data` routine.

**Reply by DanRRRR** [reply]
Thanks Tom. 
Well, i looked there, and the more i look the more scarier it appears  :(( 
This subroutine resembles and probably is a definition of hell. For AI it is probably OK but for the humans it's a thriller.   :)

Usually developers of PIC codes offer small handy programs to extract particles and field data in simple texts/binary format out of the code's compressed output files (HDF is mostly used by almost everyone which is functional analog of SDF which probably no one uses besides EPOCH)  for further processing not just for plotting like it was done with EPOCH. I asked here about this few times, do you guys plan eventually to do that?    Any means via MATLAB, Python or Fortran is OK, just it has to be not slow ( binary  output format  only allows that, specifically with the files opened with the ACCESS=STREAM option). 

Or may be you at least plan to switch to HDF5 format in future code versions? Then users could handle that themselves without the dark forest feelings like with SDF. The HDF5 has Fortran/C/Python etc small demo examples and tools to extract anything  into text/binary forms. First, some  third party programs exist which  show HDF file  structure (names, dimensions, data formats) then you just extract that needed 2D or 3D array with just one single command. Or read HDF5 file in Fortran/Python and then output in text/binary form, HDF consortium supplies all library for many languages (Intel at least)

Nothing like that exists for SDF

**Reply by Status-Mirror** [reply]
Hi Dan,

I don't see the need for further SDF development when we're planning to move away from this in the next EPOCH release anyway. Until then, this documentation may help you understand what the FORTRAN SDF readers are doing:

[SDF_documentation.pdf](https://github.com/user-attachments/files/16872544/SDF_documentation.pdf)


Cheers,
Stuart
